firedrake/projection.py

```diff
@@ -129,32 +140,61 @@ def __init__(self, v, v_out, bcs=None, solver_parameters=None, constant_jacobian
-
-            a = ufl.inner(p, q)*ufl.dx
-            L = ufl.inner(p, v)*ufl.dx
-
-            problem = vs.LinearVariationalProblem(a, L, v_out, bcs=self.bcs,
-                                                  constant_jacobian=constant_jacobian)
-
-            if solver_parameters is None:
-                solver_parameters = {}
-
-            solver_parameters.setdefault("ksp_type", "cg")
-
-            self.solver = vs.LinearVariationalSolver(problem,
-                                                     solver_parameters=solver_parameters)
+        self.method = method
+
+        if self.method == "l2":
+            if not self._same_fspace or self.bcs:
+                V = v_out.function_space()
+
+                p = ufl_expr.TestFunction(V)
+                q = ufl_expr.TrialFunction(V)
+
+                a = ufl.inner(p, q)*ufl.dx
+                L = ufl.inner(p, v)*ufl.dx
+
+                problem = vs.LinearVariationalProblem(a, L, v_out, bcs=self.bcs,
+                                                      form_compiler_parameters=form_compiler_parameters,
+                                                      constant_jacobian=constant_jacobian)
+
+                if solver_parameters is None:
+                    solver_parameters = {}
+
+                solver_parameters.setdefault("ksp_type", "cg")
+
+                self.solver = vs.LinearVariationalSolver(problem,
+                                                         solver_parameters=solver_parameters)
+        elif self.method == "average":
+            # NOTE: Any bcs on the function self.v should just work.
+            # Loop over node extent and dof extent
+            V = self.v_out.function_space()
+            shapes = (V.finat_element.space_dimension(), np.prod(V.shape))
+            self._accumulate_kernel = """
+            for (int i=0; i<%d; ++i) {
+                for (int j=0; j<%d; ++j) {
+                    vo[i][j] += v[i][j];
+                    w[i][j] += 1.0;
+            }}""" % shapes
+
+            self._w = function.Function(V)
+        else:
+            raise ValueError("Method type %s not recognized" % str(method))
 
     def project(self):
         """
         Apply the projection.
         """
-        if self._same_fspace and not self.bcs:
-            self.v_out.assign(self.v)
+        if self.method == "l2":
+            if self._same_fspace and not self.bcs:
+                self.v_out.assign(self.v)
+            else:
+                self.solver.solve()
+
         else:
-            self.solver.solve()
+            # Ensure the functions we populate into are zeroed out
+            self.v_out.dat.zero()
+            self._w.dat.zero()
+            par_loop(self._accumulate_kernel, ufl.dx, {"vo": (self.v_out, INC),
+                                                       "w": (self._w, INC),
+                                                       "v": (self.v, READ)})
+            self.v_out.dat /= self._w.dat
```

Comment:
To reduce the number of mesh traversals, you could do this instead:

```python
@utils.cached_property
def _weighting(self):
    w = Function(V)
    shapes = (V.finat_element.space_dimension(), np.prod(V.shape))
     par_loop("""
 for (int i=0; i<%d; ++i) {
                 for (int j=0; j<%d; ++j) {
                     w[i][j] += 1.0;
             }}""" % shapes, {"w": (w, op2.INC)})
 
     w.assign(1/w)

self._accumulate_kernel = "for ... for ... {vo[i][j] += v[i][j]*w[i][j];}"
```

That way you have a setup phase and then the actual averaging is just a dat zero + one further loop.